Readme.md
## Description
From 31st December 2018 Soundbox will no longer be supported and the media subscriptions will no longer be active. From 1st January 2019 there will be no inbuilt way to automatically get the media (images and videos) into Soundbox for the meetings. This script automatically copies all of the images from the Meeting Workbooks and Watchtowers in JW Library to Soundbox.

## Limitations
* The listing codes for videos are not yet correct so may appear out of sequence in Soundbox
* Media before 2018 is not copied. This is by design because the DB structure before this does not include the necessary tables for this script to operate.
* The script will copy the images from ALL meeting workbooks found in JW Library. Be aware this could cause LOTS of files and folders to appear in your Soundbox media folders.
* The Congregation Bible Study images assumes you are using jy ("Jesus - The Way") book. You must edit line 2 of the script when this changes.

## PreRequisites
* Soundbox installed and working
* JWLibrary installed and working

## Testing
Note as always you should not perform testing on a live (production) system and it is your responsibility to backup any files before deleting them and ensure correct operation of this program.

## Installation
1. Download and install the latest version of Python 3 from https://www.python.org/downloads/windows/. Make sure to tick the checkbox for "Add Python to PATH" during the installation.
2. Download and save the CopyJWMedia.py script to a suitable location on your computer such as C:\Scripts
3. By default this program will copy media from 2019 onwards. If you want to limit the copying of publications to a different year you can  edit line 1 of the CopyJWMedia.py script to specify the starting year. DO NOT SPECIFY A YEAR BEFORE 2017 OR YOU WILL GET ERROR MESSAGES AND THE SCRIPT WILL FAIL!

## Usage
1. Download the required meeting workbooks in JWLibrary.
2. Open up a command prompt (Windows key+R > cmd)
3. Type 'python' (without the quotes) followed by a space and the path to the downloaded script. E.g.
>python C:\Scripts\CopyJWMedia.py

For repeated usage you can create a scheduled task in Windows to run this python script on a regular basis.
 
## Troubleshooting
If you get the message 'python is not recognized as an internal or external command, operable program or batch file' then you have probably forgotten to add the python installation folder to the path during installation. There are 3 ways you can address this:
1. Edit the Environment Variables on your PC to add the path to your python folder.
2. Reinstall Python and remember to check the box this time.
3. Specify the full path to python when running the script. E.g.
>C:\Users\username\AppData\Local\Programs\Python\Python37\python C:\Scripts\CopyJWMedia.py

## Additional Information
A log file 'JWMediaCopier.log' will be created in the same folder as the script itself. This file will list any missing video files that have been skipped.

This script does not access the internet or in any way alter the publications provided by Watchtower Bible and Tract Society. It merely copies files from one location on your computer's hard drive to another location on your hard drive. To the best of my knowledge it does not violate any terms and conditions of JWLibrary but you must also ensure any usage is in harmony with your Bible-trained conscience.


CHANGELOG.md
## [0.6.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.6.0) - 12/01/2019
### Added
* Images for Congregation Bible Study are now copied.

## [0.5.2](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.5.2) - 12/01/2019
### Fixed
* Watchtower images weren't copying if file path was over 260 characters.

## [0.5.1](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.5.1) - 12/12/2018
### Fixed
* Introduction to Bible book videos weren't copying.

## [0.5.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.5.0) - 15/10/2018
### Added
* Missing files are now logged.

## [0.4.1](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.4.1) - 01/10/2018
### Fixed
* Missing video files are now skipped.

## [0.4.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.4.0) - 27/09/2018
### Added
* Videos now copy

## [0.3.1](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.3.1) - 23/09/2018
### Fixed
* Strip question marks from file names

## [0.3.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.3.0) - 22/09/2018
### Added
* Images from the Watchtower are now copied to Soundbox.
### Changed
* Publications before 2017 are ignored.

## [0.2.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.2.0) - 19/09/2018
### Changed
* Images are now copied to the correct week and prefixed with the listing codes.
* Only the high resolution images are copied to Soundbox.

## [0.1.0](https://github.com/Caltor/JWMediaCopier/releases/tag/v0.1.0) - 14/09/2018
### Added
* Initial Release
* Test
4e663355beab86479f09bc176749d8b81962ae50


TestSqlite.py
import sqlite3

dbpath = "C:\\Users\\darren\\AppData\\Local\\Packages\\WatchtowerBibleandTractSo.45909CDBADF3C_5rz59y55nfz3e\\LocalState\\Publications\\mwb_E_201809\\mwb_E_201809.db"
year = "2018"
month = "09"

conn = sqlite3.connect(dbpath)
conn.row_factory = sqlite3.Row # allows accessing columns using column name - see https://docs.python.org/2/library/sqlite3.html#row-objects
c = conn.cursor()
for row in c.execute("SELECT * FROM Document ORDER BY DocumentId"):
    # print(row['Title'])
    if row['Class'] == '106':
        # This is a new week
        week = row['Title']
        print("Week: ", week)
        split_week = week.split()
        print('Month', split_week[0])
        print('Dates', split_week[1])   #get first character
        first_and_last = split_week[1].split('-')
        date = first_and_last[0]
        print("First Date:", date)
        targetpath = year + "-" + month + "-" + str(date).zfill(2)
        print("targetpath", targetpath)

    if row['Class'] in ['21','107','10']:
        # Treasures from God's word or Living as Christians
        document_id = row['DocumentId']
        #print("Document: ", document_id)

        # Get all of the multimedia records for this document
        d = conn.cursor()
        t_doc = (document_id, )
        for row2 in d.execute("SELECT DocumentMultimedia.MultimediaId, Label, Filepath FROM DocumentMultimedia JOIN Multimedia ON DocumentMultimedia.MultimediaId = Multimedia.MultimediaId WHERE DocumentId = ? AND CategoryType = 8", t_doc):
            print(row2['MultimediaId'])
            sourcefile = row2['Filepath']
            targetfile = row2['Label'] + ".jpg"
            print("Source file:", sourcefile)
            print("Target file:", targetfile)

conn.close()


CopyJWMedia.py
starting_year = 2019
study_publication = "jy"
study_publication_full = study_publication + "_E"
jwlibrary_package = "WatchtowerBibleandTractSo.45909CDBADF3C_5rz59y55nfz3e"

## TODO:
## * Split code into functions rather than one big script
## * Allow user to copy only Watchtower or only Meeting Workbooks using command line switches

import os, calendar, shutil, time, sqlite3, logging
from datetime import date, timedelta

def get_filtered_folders(search_string, array):
    return list(filter(lambda x: search_string in x, array))

def get_year_month_from_folder(folder_name):
    year = folder_name[-6:-2]
    month = folder_name[-2:]
    return (year, month)

def get_db_connection(dbpath):
    conn = sqlite3.connect(dbpath)
    conn.row_factory = sqlite3.Row # allows accessing columns using column name - see https://docs.python.org/2/library/sqlite3.html#row-objects
    return conn

def get_documents(conn):
    c = conn.cursor()
    return c.execute("SELECT * FROM Document ORDER BY DocumentId")

def get_documents_by_meps_document_id(conn, documentid):
    c = conn.cursor()
    return c.execute("SELECT * FROM Document WHERE MepsDocumentId = ?", (str(documentid),) )

def get_document_multimedia(conn, documentid):
    c = conn.cursor()
    return c.execute("SELECT MultimediaId FROM DocumentMultimedia WHERE DocumentId = ?", (str(documentid),) )

def get_document_multimedia_info(conn, documentid):
    c = conn.cursor()
    return c.execute("SELECT DocumentMultimedia.MultimediaId, Label, Filepath FROM DocumentMultimedia JOIN Multimedia ON DocumentMultimedia.MultimediaId = Multimedia.MultimediaId WHERE CategoryType = 8 AND DocumentId = ?", (str(documentid),))  #need the extra comma as we pass in a tuple

def get_media_keys(media_conn, issuetag, track):
    c = media_conn.cursor()
    return c.execute("SELECT * FROM MediaKey WHERE IssueTagNumber = ? AND Track = ?", (issuetag, track))
 
def get_media_key(media_conn, issuetagnumber, track):
    c = media_conn.cursor()
    c.execute("SELECT MediaKeyId FROM MediaKey WHERE IssueTagNumber = ? AND Track = ?", (issuetagnumber, track))
    row = c.fetchone()
    if row is None:
        return
    return row['MediaKeyId']

def get_first_date_wt(conn):
    c = conn.cursor()
    c.execute("SELECT FirstDatedTextDateOffset FROM Publication")
    row = c.fetchone()
    return int2date(row[0])

def get_meps_document_ids(conn, documentid):
    c = conn.cursor()
    c.execute("select RefMepsDocumentId from Extract inner join RefPublication on Extract.RefPublicationId = RefPublication.RefPublicationId inner join DocumentExtract on Extract.ExtractId = DocumentExtract.ExtractId where DocumentExtract.DocumentId = ? and RefPublication.RootSymbol = ?", (str(documentid), study_publication))
    return c

def get_multimedia_tag(conn, multimedia_id):
    c = conn.cursor()
    c.execute("SELECT KeySymbol, IssueTagNumber, Track FROM Multimedia WHERE MultimediaId = ?", (str(multimedia_id),) )
    row = c.fetchone() 
    return (row['KeySymbol'], row['IssueTagNumber'], row['Track'])

def get_video_details(media_conn, media_key):
    c = media_conn.cursor()
    c.execute("SELECT Title, Filepath FROM Video WHERE MediaKeyId = ?", (media_key,) )
    return c.fetchone()
    
def copyfile_nooverwrite(source, target):
    if not os.path.exists(target):
        shutil.copy2(source, target)

def sanitise_filename(filename):
    for char in "<>:\"/\\|?*":
        filename = filename.replace(char, '')
    return filename

def int2date(argdate: int) -> date:
    """
    If you have date as an integer, use this method to obtain a datetime.date object.

    Parameters
    ----------
    argdate : int
      Date as a regular integer value (example: 20160618)

    Returns
    -------
    dateandtime.date
      A date object which corresponds to the given value `argdate`.
    """
    year = int(argdate / 10000)
    month = int((argdate % 10000) / 100)
    day = int(argdate % 100)

    return date(year, month, day)
                    
## Main program ##
print("Copying images from JW Library Meeting Workbooks to Soundbox...")

meeting_parts = {
    '10': '3',  # Living as Christians
    '21': '1',  # Treasures from God's Word
    '107': '3' # Living as Christians
    }

start = time.time()

## setup the logger
logging.basicConfig(filename='JWMediaCopier.log', level=logging.WARNING, format='%(asctime)s %(levelname)s %(message)s', filemode='w')  ##overwrites the log each time

## Open Media Catalog
media_catalog_path = os.path.join(os.getenv("LOCALAPPDATA"), "packages", jwlibrary_package, "LocalState", "Data", "mediaCollection.db")
media_conn = get_db_connection(media_catalog_path)

# By using os.path.join() instead of backslashes we make this script cross-platform compatible. You know for when we get JWlibrary and Soundbox for Linux and Mac! ;) teehee...
targetpath_base = os.path.join(os.getenv("ProgramData"), "SoundBox", "MediaCalendar")
path = os.path.join(os.getenv("LOCALAPPDATA"), "packages", jwlibrary_package, "LocalState", "Publications")

## Open Congregation Bible Study catalog
book_study_path = os.path.join(path, study_publication_full)
book_study_database = os.path.join(book_study_path, study_publication_full + ".db")
book_study_conn = get_db_connection(book_study_database)

array = os.listdir(path)


print("\r\nMeeting Workbooks:")
filtered_folders = get_filtered_folders("mwb_E_", array)
for source_folder in filtered_folders:
    year, month = get_year_month_from_folder(source_folder)
    
    ## Ignore years before 2017
    if int(year) < starting_year:
        continue

    print("Copying " + source_folder)
        
    source_path = os.path.join(path, source_folder)
    dbpath = os.path.join(source_path, source_folder+".db")

    conn = get_db_connection(dbpath)
    c = conn.cursor()
    for row in c.execute("SELECT * FROM Document ORDER BY DocumentId"):
        row_class = row['Class']
        if row_class == '106':
            # This is a new week
            documentid = row['DocumentId']
            week = row['Title']
            split_week = week.split('-')    #splits into 'from' and 'to' sections
            if len(split_week) == 1:
                split_week = week.split('–')

            from_date = split_week[0].split()  # splits into month and date
            first_date = from_date[1]
            target_folder = year + "-" + month + "-" + str(first_date).zfill(2)
            targetpath = os.path.join(targetpath_base, year, target_folder)
            print("Writing files to", targetpath)
            if not os.path.exists(targetpath):
                os.makedirs(targetpath)

            ## Get the Congregation Bible Study images
            document_ids = get_meps_document_ids(conn, documentid)
            for row in document_ids:
                ## Get jy.db -> Document -> DocumentMultimedia -> Multimedia etc
                counter=0
                for book_study_doc in get_documents_by_meps_document_id(book_study_conn, row['RefMepsDocumentId']):
                    counter += 10
                    for media in get_document_multimedia_info(book_study_conn, book_study_doc["DocumentId"]):
                        book_study_source_file = media['Filepath']
                        target_file_name = "M3-" + str(counter).zfill(3) + " " + media['Label'].replace('?', '')
                        target_file_path = os.path.join(targetpath, target_file_name)[:255] + ".jpg"
                        if not os.path.exists(target_file_path):
                            source_file_path = os.path.join(book_study_path, book_study_source_file)
                            shutil.copyfile(source_file_path, target_file_path)
                
                
            ## Get the Videos!!!
            document_multimedia_records = get_document_multimedia(conn, documentid)
            counter = 0
            for document_multimedia in document_multimedia_records:
                multimedia_id = document_multimedia['MultimediaId']
                keysymbol, issuetag, track = get_multimedia_tag(conn, multimedia_id)
                if keysymbol == 'nwtsv' or issuetag > 0:
                    media_key = get_media_key(media_conn, issuetag, track)
                    row = get_video_details(media_conn, media_key)
                    if row:
                        source_file_path = row['Filepath']
                        title = row['Title']
                        valid_file_name = sanitise_filename(title)
                        meeting_part = '1'
                        counter += 10
                        target_file_name = "M" + meeting_part + "-" + str(counter).zfill(3) + " " + valid_file_name + ".mp4"
                        target_file_path = os.path.join(targetpath, target_file_name)
                        if os.path.exists(source_file_path):
                            if not os.path.exists(target_file_path):
                                shutil.copyfile(source_file_path, target_file_path)
                        else:
                            warning_message = "File " + source_file_path + " was not found - skipped"
                            logging.warning(warning_message)
                            print("Warning: " + warning_message)


        if row_class in ['21','107','10']:
            # Treasures from God's word or Living as Christians
            document_id = row['DocumentId']

            # Get all of the multimedia records for this document
            d = conn.cursor()
            t_doc = (document_id, )
            counter = 0
            for row2 in d.execute("SELECT DocumentMultimedia.MultimediaId, Label, Filepath FROM DocumentMultimedia JOIN Multimedia ON DocumentMultimedia.MultimediaId = Multimedia.MultimediaId WHERE DocumentId = ? AND CategoryType = 8", t_doc):
                counter += 10
                sourcefile = row2['Filepath']
                meeting_part = meeting_parts[row_class]
                target_file_name = "M" + meeting_part + "-" + str(counter).zfill(3) + " " + row2['Label'].replace('?','')
                target_file_path = os.path.join(targetpath, target_file_name)[:255] + ".jpg"
                if not os.path.exists(target_file_path):
                    source_file_path = os.path.join(source_path, sourcefile)
                    shutil.copyfile(source_file_path, target_file_path)

    conn.close()


## WatchTower
print("\r\nWatchtowers:")
filtered_folders = get_filtered_folders("w_E_", array)
for source_folder in filtered_folders:
    dbpath = os.path.join(path, source_folder, source_folder+".db")
    conn = get_db_connection(dbpath)
    study_date = get_first_date_wt(conn)
    study_date += timedelta(days=6) # Change w/c date from Monday to Sunday

    ## Create target folder
    year = str(study_date.year)

    ## Ignore years before 2017
    if int(year) < starting_year:
        continue
    
    print("Copying "+source_folder)

    for document in get_documents(conn):
        document_class = document['Class']
        if document_class == '40':
            folder_name = str(study_date)
            targetpath = os.path.join(targetpath_base, year, folder_name)
            print("Writing files to", targetpath)
            if not os.path.exists(targetpath):
                os.makedirs(targetpath)
                
            counter = 0
            images = get_document_multimedia_info(conn, document['DocumentId'])
            for image in images:
                counter += 10
                source_file_path = os.path.join(path, source_folder, image['Filepath'])
                target_file_name = "W2-" + str(counter).zfill(3) + " " + image['Label'].replace('?','')
                target_file_path = os.path.join(targetpath, target_file_name)[:255] + ".jpg"
                copyfile_nooverwrite(source_file_path, target_file_path)

            study_date += timedelta(days=7) #Increment the week
            
    conn.close()

media_conn.close()

elapsed = str(time.time() - start)
print("\r\nFinished")
print("Time taken: " + elapsed + " seconds")


